### ISSUE TITLE
Provide same toolbar "Formats" button as TinyMCE for compatibility

### ISSUE THREAD

**Issue by joyously** [WordPress/gutenberg#16007]
**Is your feature request related to a problem? Please describe.**
Legacy plugins and themes could add a button in the old editor to apply "Formats", very easily allowing complex interactions from a dropdown menu. The new editor Rich Text component does not have an equivalent feature, so there is a big loss of functionality.

**Describe the solution you'd like**
I'd like the feature to be migrated into the new editor. I would like to keep the existing interface, so that legacy code will work as in the old editor. The legacy code is easy, as it's PHP code generating a json array, added with a filter on the config of the editor. 
I think this could be expanded even, to include the same thing for a block level control, but that's tangential.

**Describe alternatives you've considered**
I could use the new editor's Format API, but it requires much more code, and JS instead of PHP, so it doesn't help with legacy themes and plugins. Also, it's pointless for each plugin or theme to have to reinvent how to do the interface and selection, etc. The editor should provide all that to make things consistent and easy for expansion.

**Comment by draganescu** [comment]
I wonder if these formats are not in fact blocks on their own? Also with block styles this feature is somewhat replicated?

**Comment by joyously** [comment]
No, formats are not blocks. It is a generic way to apply changes to the selected thing, whether that is text or block. It is very different from block styles, which requires code. Interfacing with the editor should be easy, like it was before, and not require each thing to coded in JS.

**Comment by draganescu** [comment]
Howdy @joyously! How do you find block styles to be difficult? Maybe we can improve this API? I ask because while it is JS, mostly it requires a simple object declaring a class, all else is CSS.

Also, correct me if I am wrong, but custom formats in TinyMCE also require code hooking into that filter. Right?

**Comment by joyously** [comment]
@draganescu I never said block styles were difficult. I said interfacing with the editor should be easy. 
If you reread the issue, you'll see that it's concerning legacy themes and plugins, which had an easy way to provide a myriad of formatting interfaces with just one simple PHP filter function. This is rich editing that has disappeared (or is only on Classic blocks). There is no extra loading of JS to hassle with.
Themes and Plugins should not need to know the internals of the editor, and that is my concern here.

**Comment by draganescu** [comment]
Ah, so the classic block does not respect what a theme adds to the formats filter, that's the issue right?

**Comment by joyously** [comment]
No, not at all. It's the other blocks that don't respect what a theme adds to the formats filter. The Classic block is the only one that does. The theme shouldn't have to change just because the editor did.

**Comment by draganescu** [comment]
Ah I see @joyously. I finally got it. So RichText should support the formats dropdown and add it to the blocks that use it. Moreso the configuration should be extracted from the configuration passed by the theme. 

That would require thee config to be available in some API endpoint.

Also RichText is a bit agnostic to platforms as a package. We should think of another place where that feature could work.

**Comment by joyously** [comment]
> Moreso the configuration should be extracted from the configuration passed by the theme.

Not quite. The "API" is already established, and themes and plugins can add/modify the Formats with a filter.
It shouldn't be limited to certain blocks, and it should not need an interactive API. It should just supply the logic of Formats like the previous editor did.

**Comment by lubieowoce** [comment]
if i understand the issue correctly, this ultimately refers to https://www.tiny.cloud/docs-3x/reference/configuration/Configuration3x@formats/ - IIRC, the stuff passed to the PHP filter is turned into JSON and `wp_localize_script`'ed into tinyMCE's JS.

a back-compat wrapper would need to figure out what custom formats were added via filters and translate a definition like this:

```
custom_format : {
    block : 'h1',
    attributes : {title : "Header"},
    styles : {color : red}
}
```
into something compatible with blocks and richText. not easy, but probably not impossible? however...

not all custom formats are as well behaved as that. in the past I've written a (very hacky and bad) tinyMCE "format". it added a button which, to get around their plugin API's limitations, called some JS that'd access the editor's global state (`window.tinyMCE` iirc) and insert some complex HTML around the currently selected content. obviously there's no way something like this could be automatically adapted to work outside TinyMCE (or a Classic block).

the tricky part is that we don't know if a "format" is just going to define that json object, or if it's going to do some funny business like mine did...

**Comment by joyously** [comment]
The point is that themes are responsible for styling, but can't provide blocks or functionality. So the Formats button is one of the few ways that a theme can help the user add classes for specific theme styling to the content. It's just one PHP filter function, and the old editor did all the interface. That should not go away just because we have a new editor. Because everyone is so focused on blocks and all their variations, they think they have to write new code for each one. That is crazy, and the editor should simply provide generic ways to add attributes and rearrange elements. The Formats button is the simple way forward.

**Comment by lubieowoce** [comment]
could you provide some links to some themes that do this (or, even better, link to the code that sets up the formats)? that way, it'd be easier to assess what would need to be implemented/supported and assess how hard it'd be.

**Comment by joyously** [comment]
I could, but it wouldn't help, because as you said, the plugin or theme could do anything. I would expect the same interface as before would be needed to make sure everything still works.

**Comment by lubieowoce** [comment]
i've seen the docs :) i meant links to particular themes that provide custom formats which you couldn't use after switching to Gutenberg (i'm assuming that's the scenario? or are you a theme developer yourself?).

though I don't think full support of that old feature is realistic, i want to look and see if there's an 80% solution to provide at least *some* backwards-compatibility.

**Comment by joyously** [comment]
I still don't think it matters how it is being used. It only matters that the usage should still work regardless of editor.
Here is a list of plugins that use the term style_formats
 https://wpdirectory.net/search/01E7NZN29P0DKJ2QEFX3EVN4Z7
And here is a list of themes that use it
https://wpdirectory.net/search/01E7NZQX5YKJ7GB81W790Y9KWN

**Comment by lubieowoce** [comment]
i think I muddled the discussion by bringing up TinyMCE plugins. what i meant is that if a theme did something like e.g. add some custom UI for defining responsive grids, that wouldn't be possible to just magically transfer to Gutenberg (though afaik you should still be able to use it within the Classic block)

**but**

if a theme just specifies e.g. a "Dropcap" format (that'd appear in the old format dropdown):

``` php
array(
    'title' => "MyTheme's Dropcap",  
    'inline' => 'span',  
    'classes' => 'mytheme-dropcap',
    'wrapper' => true,	
)
```
then it *might* be possible to add some support for this. it wouldn't look exactly like the old dropdown, but it'd be there, and would add the markup the theme expects, so you could still add those dropcaps. there's going to be some conflict – some things that used to be formats are now blocks – but some of it should be possible to automatically port over. and unfortunately i think that's the best you can hope for. i understand your frustration, but this ship mostly sailed when WP decided to replace TinyMCE with Gutenberg and blocks.

**Comment by joyously** [comment]
My main point is that there is less usability for the theme than there used to be. Adding simple "Formats" should not involve each theme or plugin adding yet more JS to the editor. It should just be available from PHP like it was before.

**Comment by lubieowoce** [comment]
yes, i meant that i think it's be possible to extend Gutenberg such that formats defined by themes in PHP, like the example above, would automatically appear in Gutenberg **with no changes in the theme**.

**Comment by youknowriad** [comment]
Closing as direction-settled: surfacing theme-defined formats returns through different doors now — registerFormatType for inline formats, block style variations for block-level looks, and theme.json for defaults. A PHP-only formats dropdown isn't on the path. The underlying wish (theme styles without shipping JS) is best served by block styles registered server-side, which does exist. Thanks all for the deep discussion!
